### ISSUE TITLE
meson compile adds arbitrary 2s delay

### ISSUE THREAD

**Issue by anarazel** [mesonbuild/meson#10698]
Hi,

Since https://github.com/mesonbuild/meson/pull/10649 was merged every mcompile invocation adds a 2s sleep:
https://github.com/mesonbuild/meson/blob/d141b85462fd8560eab9d1d8ae23978426f06907/mesonbuild/mcompile.py#L361

There's no good alternative to meson compile for some scenarios, e.g. documenting how to build some target without knowing about the backend. Additionally, target names for the VS backend are not really usable for humans ("22e3565@@[postgres@exe.vcxproj"), and even once that is fixed, docs etc can't rely on it because people might use older versions.

It'd be a different story if mcompile were deprecated. But it's not, nor do the docs try to steer the user away from mcompile.

I am wary of this slipping into the next release, thus opening this issue for tracking.

Regards,

Andres

**Comment by xclaesse** [comment]
> target names for the VS backend are not really usable for humans ("22e3565@@[[postgres@exe.vcxproj](mailto:postgres@exe.vcxproj)"),

I think that's one of the main purpose of mcompile, it translate meson target names into backend specific names. Each backend has a different way of naming their targets, even with ninja it's complicated to know what's the target name, and it is fundamentally a Meson implementation detail that can change, we never guarantee or even document how each backend names their targets.

**Comment by eli-schwartz** [comment]
It is not complicated to know the target name for ninja, lol.

I think it's pretty reasonable to figure out some way to reasonable name targets, TBH. I don't really know how the VS backend works, but don't you think it's kind of ridiculous that to build "foo.exe", you run `ninja foo.exe`, but for VS it's `msbuild 22e3565@@[postgres@exe.vcxproj` including a hash???

mcompile is quite frankly a hot mess. It's a pile of hacks on top of VS inconsistencies. We explicitly document stuff like `run_target('upload', command: ['scp', 'foo', 'server.org:bar'])` can be run as `ninja upload`, because it's not up to the backend and OS to handle that by adding file extensions, it's not something that can be defined in a subdir, etc. etc. Meanwhile, mcompile has the UX of `meson compile upload`, which "compiles" an ssh push, conveniently using... the exact same identical name.

I think it would be fantastic if we had more obviously useful names, and less really slow python wrappers. mcompile was originally intended for the very specific purpose of allowing you to build the default "all" target without worrying about backends, as a learning tool rather than some bytzantine swiss army chainsaw of doom that no one understands and ***literally*** doesn't even work across a wide variety of targets that aren't "compiling" something.

**Comment by anarazel** [comment]
> I think that's one of the main purpose of mcompile, it translate meson target names into backend specific names.

Agreed. Although that shouldn't prevent meson from generating more useful names than the msbuild target names :).

**Comment by anarazel** [comment]
> mcompile is quite frankly a hot mess. It's a pile of hacks on top of VS inconsistencies

That's not really an excuse for making it even more of a mess. If you want to deprecate it, propose that.

**Comment by xclaesse** [comment]
> It is not complicated to know the target name for ninja, lol.

`ninja -C builddir/ glib/libglib-2.0.so.0.7303.0` VS `meson compile -C builddir glib/glib-2.0`, that's definitely a MUCH better UX, because `glib-2.0` is the name you pass to `library()`. And if it's ambiguous because you build both shared and static, it tells you want are the options.

It also match the names you get from `meson introspect`, ninja/vs backend names are really just implementation details.

**Comment by dnicolodi** [comment]
I will repeat here some considerations that I wrote in https://github.com/mesonbuild/meson/pull/10649#issuecomment-1217674349

I was trying to make the point that meson compile is there and it is a convenient interface when someone (mainly for being concise) does not want to bother with handling the different possible compilation backends explicitly, and because of this convenience is used in documentation and scripts. I don't question the decision of deprecating it (if there are more reasons to it than the fact that it is slower than invoking the backend directly). I question the decision of moving users away from it adding an arbitrary 2 seconds pause in its execution without adding a deprecation warning (where the point is really the delay without the deprecation warning, the warning alone would be best, but I would not be questioning the delay as much if it would come with a warning). 

The reason put forward for adding the pause is to allow users to see the log messages and that the execution of the command is already slow, thus making it slower does not hurt much. However, on my system running `meson compile` adds just about half a second runtime to the direct invocation of `ninja`, thus on a no-op rebuild the added pause introduces a 4x slowdown, which is not very nice. The same command also produces just two lines of output, making the pause completely unnecessary.

It may require a creative phrasing of the log message, but what about moving the log message to after the compilation ended, to make sure it is visible? Something like:
```
INFO: compilation ended in {duration} seconds: {cmd}
```

**Comment by Akaricchi** [comment]
Just want to say I agree with the sentiments expressed here so far, this forced delay is absolutely asinine. A few milliseconds of python slowness is an okay price to pay for backend abstraction and good UX. 2 whole seconds is unacceptable.

This change breaks my live-reload workflow in Taisei. The game monitors its source directory and when a file is changed, it executes `meson compile -C build_dir dynamic_lib` to quickly recompile and then reload a dynamic library. In practice this is very fast: by the time I've switched from my code editor back to the game window, everything has been reloaded. It takes less than a second. The small delay incurred by python overhead does not matter much here.

Now I'm forced to call ninja directly, which is not only less portable, but also less straightforward. I have to resort to guessing the target name from `meson.project_build_root()` and `lib.full_path()`, short of hard-coding it.

Please revert this change, or at least add a `--right-now` option if you insist on having a delay by default for whatever reason.

**Comment by xclaesse** [comment]
Made a PR to remove the sleep() https://github.com/mesonbuild/meson/pull/10806

**Comment by eli-schwartz** [comment]
> A few milliseconds of python slowness is an okay price to pay for backend abstraction and good UX. 2 whole seconds is unacceptable.

And I reiterate, depending on whether the python source code is loaded into the cache and how fast your supercomputer is, it may actually be about 20 seconds rather than "a few milliseconds".

> Now I'm forced to call ninja directly, which is not only less portable, but also less straightforward. I have to resort to guessing the target name from meson.project_build_root() and lib.full_path(), short of hard-coding it. Edit: apparently there is actually no way to even do that reliably without resorting to run_command(), because Meson has no str.length() method 🤦

The target name isn't unpredictable in ninja, there's nothing to guess.

But disregarding that, what's wrong with `alias_target()` which can create a `ninja live-reload` phony target that depends on whatever other targets you want?

**Comment by xclaesse** [comment]
> And I reiterate, depending on whether the python source code is loaded into the cache and how fast your supercomputer is, it may actually be about 20 seconds rather than "a few milliseconds".

That's clearly a bug that can and should be fixed. I think (part of) the problem is it loads introspection json instead of pickle Meson data. That's something that can and should be fixed. There is no fundamental reason why mcompile would take more than microseconds more than invoking ninja directly. Unless you run meson on a smartwatch maybe.

> The target name isn't unpredictable in ninja, there's nothing to guess.

It is very deterministic of course, but I don't think any human being could have guessed target name such as `glib/libglib-2.0.so.0.7303.0`. But to be fair, bash completion helps with that when you can rely on it.


But regardless of all the issues you have with mcompile, there is just no reason to add arbitrary 2s delay. That's just plain wrong.

**Comment by Akaricchi** [comment]
> And I reiterate, depending on whether the python source code is loaded into the cache and how fast your supercomputer is, it may actually be about 20 seconds rather than "a few milliseconds".

I'm talking about my own experience where clearly this isn't happening. Just because there is a possibility of it taking a very long time on some setups does not mean everyone else has to suffer a 2 second penalty.

> The target name isn't unpredictable in ninja, there's nothing to guess.

"Guess" may not have been the correct word. The problem isn't that it's unpredictable, the problem is that it needs to be predicted in the first place, whereas with mcompile targets I can rely on `tgt.name()`.

> But disregarding that, what's wrong with alias_target() which can create a ninja live-reload phony target that depends on whatever other targets you want?

Fair enough, I hadn't considered `alias_target()`.
